tfhe/examples/regex_engine/parser.rs

```diff
@@ -0,0 +1,685 @@
+            }
+            let mut re_xs = vec![];
+            if sof.is_some() {
+                re_xs.push(RegExpr::SOF);
+            }
+            re_xs.push(re);
+            if eof.is_some() {
+                re_xs.push(RegExpr::EOF);
+            }
+            RegExpr::Seq { re_xs }
+        }),
+        optional(byte(b'i')),
+    ))
+        .map(|(re, case_insensitive)| {
+            if case_insensitive.is_some() {
+                re.case_insensitive()
+            } else {
+                re
+            }
+        })
+        .parse(pattern.as_bytes())?;
+    if !unparsed.is_empty() {
+        return Err(anyhow!(
+            "failed to parse regular expression, unexpected token at start of: {}",
+            std::str::from_utf8(unparsed).unwrap()
+        ));
+    }
+
+    Ok(parsed)
+}
+
+// based on grammar from: https://matt.might.net/articles/parsing-regex-with-recursive-descent/
+//
+//  <regex> ::= <term> '|' <regex>
+//           |  <term>
+//
+//  <term> ::= { <factor> }
+//
+//  <factor> ::= <base> { '*' }
+//
+//  <base> ::= <char>
+//          |  '\' <char>
+//          |  '(' <regex> ')'
+
+parser! {
+    fn regex[Input]()(Input) -> RegExpr
+        where [Input: Stream<Token = u8>]
+        {
+            regex_()
+        }
+}
+
+fn regex_<Input>() -> impl Parser<Input, Output = RegExpr>
+where
+    Input: Stream<Token = u8>,
+    Input::Error: ParseError<Input::Token, Input::Range, Input::Position>,
+{
+    choice((
+        attempt(
+            (term(), byte(b'|'), regex()).map(|(l_re, _, r_re)| RegExpr::Either {
+                l_re: Box::new(l_re),
+                r_re: Box::new(r_re),
+            }),
+        ),
+        term(),
+    ))
+}
+
+fn term<Input>() -> impl Parser<Input, Output = RegExpr>
+where
+    Input: Stream<Token = u8>,
+    Input::Error: ParseError<Input::Token, Input::Range, Input::Position>,
+{
+    many(factor()).map(|re_xs: Vec<RegExpr>| {
+        if re_xs.len() == 1 {
+            re_xs[0].clone()
+        } else {
+            RegExpr::Seq { re_xs }
+        }
+    })
+}
+
+fn factor<Input>() -> impl Parser<Input, Output = RegExpr>
+where
+    Input: Stream<Token = u8>,
+    Input::Error: ParseError<Input::Token, Input::Range, Input::Position>,
+{
+    choice((
+        attempt((atom(), byte(b'?'))).map(|(re, _)| RegExpr::Optional {
+            opt_re: Box::new(re),
+        }),
+        attempt(repeated()),
+        atom(),
+    ))
+}
+
+const NON_ESCAPABLE_SYMBOLS: [u8; 14] = [
+    b'&', b';', b':', b',', b'`', b'~', b'-', b'_', b'!', b'@', b'#', b'%', b'\'', b'\"',
+];
+
+fn atom<Input>() -> impl Parser<Input, Output = RegExpr>
+where
+    Input: Stream<Token = u8>,
+    Input::Error: ParseError<Input::Token, Input::Range, Input::Position>,
+{
+    choice((
+        byte(b'.').map(|_| RegExpr::AnyChar),
+        attempt(byte(b'\\').with(parser::token::any())).map(|c| RegExpr::Char { c }),
+        choice((byte::letter(), parser::token::one_of(NON_ESCAPABLE_SYMBOLS)))
```

Comment:
instead of `byte::letter()` I think it would be better to use `byte::alpha_num` to be able to do regex with numbers, the range rule (and maybe others) would also  need to be update
(and add tests)

Reply:
Other comments addressed. This comment I will address later today after work, as it requires a bit work.